### PR TITLE
Kafka: AWS IAM role auth support

### PR DESCRIPTION
**Pull request by eunicecompra**
Raised in dapr /components-contrib /pull /3310

### CODE REVIEW

File: common/component/kafka/metadata.go [hunk 1]
```diff
@@ -83,6 +83,7 @@ type KafkaMetadata struct {
 	AWSAccessKey         string `mapstructure:"awsAccessKey"`
 	AWSSecretKey         string `mapstructure:"awsSecretKey"`
 	AWSSessionToken      string `mapstructure:"awsSessionToken"`
+	AWSIamRoleArn        string `mapstructure:"awsIamRoleArn"`
```
Comment (chrispidcock) [inline common/component/kafka/metadata.go:86]: Do you want to make the context timeout configurable? I think that would make sense, and should be added somewhere here I would imagine.

File: common/component/kafka/auth.go [hunk 2]
```diff
@@ -128,6 +131,13 @@ func (m *mskAccessTokenProvider) Token() (*sarama.AccessToken, error) {
 			}, nil
 		}))
 		return &sarama.AccessToken{Token: token}, err
+	} else if m.awsIamRoleArn != "" {
+		// Using a different context since the context from init may have been cancelled when token is refreshed
+		ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
```
Comment (chrispidcock) [inline common/component/kafka/auth.go]: Configuring this timeout may be important for some users IMO

Comment (eunicecompra) [inline common/component/kafka/auth.go]: Context implementation will be updated based on the outcome of https://github.com/dapr/components-contrib/pull/3305

File: common/component/kafka/metadata.go (hunk 1 above)
Comment (chrispidcock) [reply common/component/kafka/metadata.go:86]: I think you would want to handle your changes to `requried` by checking for which metadata values were provided.

https://github.com/eunicecompra/components-contrib/blob/1abad9324e0e42014d7e6d8ef7aff4578af5c975/common/component/kafka/metadata.go#L237C1-L241C56

Example:

```golang
	case awsIAMAuthType:
		if m.AWSRegion == "" {
			return nil, errors.New("missing AWS region property 'awsRegion' for authType 'awsiam'")
		}
                if m.AWSIamRoleArn == "" && m.AWSSecretKey  == "" && m.AWSAccessKey  == "" {
                        return nil, errors.New("missing required properties for authType 'awsiam'")
                }
		k.logger.Debug("Configuring AWS IAM authentication.")
```

File: pubsub/kafka/metadata.yaml [hunk 3]
```diff
@@ -170,6 +168,11 @@ authenticationProfiles:
         description: |
           AWS session token to use. A session token is only required if you are using\ntemporary security credentials.
         example: '"TOKEN"'
+      - name: awsIamRoleArn
+        type: string
+        description: |
```
Comment (chrispidcock) [inline pubsub/kafka/metadata.yaml:176]: I think this description is OK, but I think the [overarching description](https://github.com/eunicecompra/components-contrib/pull/1/files#diff-7cbcd10673dca6e24c7064f4d1a23d9be6c876891114f24376b18e1015b61987L137) is bad and should be updated. This authType should support `creds` or `arn` as config, so I feel those details could be better described.

I feel like this description is not very good and could be improved: https://github.com/eunicecompra/components-contrib/pull/1/files#diff-7cbcd10673dca6e24c7064f4d1a23d9be6c876891114f24376b18e1015b61987L137

**Review by chrispidcock — COMMENTED** [review]
Looking Good @eunicecompra @rea-eunice!
I have tried to explore some of the code, which is thankfully set out in a familiar way. I added some links in comments to other parts of the code that you may want to update.

File: common/component/kafka/metadata.go (hunk 1 above)
Comment (eunicecompra) [reply common/component/kafka/metadata.go:86]: > Do you want to make the context timeout configurable?

That's a good point. However, there's another PR that fixes a context issue of the first implementation of AWS IAM which leans on defining the context at the parent context level. https://github.com/dapr/components-contrib/pull/3305/files#r1443171780

My timeout implementation could change based on the outcome of the PR.

Comment (eunicecompra) [reply common/component/kafka/metadata.go:86]: > I think you would want to handle your changes to requried by checking for which metadata values were provided.

Yes, good catch!!

File: pubsub/kafka/metadata.yaml (hunk 3 above)
Comment (eunicecompra) [reply pubsub/kafka/metadata.yaml:176]: Fair call. I will propose a change to the existing desc

**Pull request closed without merging**